File: stacks/node/base/code/testing-utilities.tpl.js
/**
 * File: testing-utilities.tpl.js
 * Purpose: Template for unknown implementation
 * Generated for: {{PROJECT_NAME}}
 */

#!/usr/bin/env node
/**
 * Node.js Testing Utilities Template
 * Purpose: Reusable testing utilities and helpers for Node.js projects
 * Usage: Import and adapt for consistent testing patterns across the application
 */

const fs = require('fs').promises;
const path = require('path');
const { EventEmitter } = require('events');

/**
 * Test data manager for creating and managing test data
 */
class TestDataManager {
    constructor() {
        this.data = this._createSampleData();
        this.tempFiles = [];
        this.tempDirs = [];
    }

    /**
     * Create sample test data
     */
    _createSampleData() {
        return {
            users: [
                {
                    id: 1,
                    username: 'testuser1',
                    email: 'test1@example.com',
                    isActive: true,
                    createdAt: new Date('2023-01-01T00:00:00Z')
                },
                {
                    id: 2,
                    username: 'testuser2',
                    email: 'test2@example.com',
                    isActive: false,
                    createdAt: new Date('2023-01-02T00:00:00Z')
                }
            ],
            posts: [
                {
                    id: 1,
                    userId: 1,
                    title: 'Test Post 1',
                    content: 'This is test content',
                    published: true,
                    createdAt: new Date('2023-01-01T12:00:00Z')
                },
                {
                    id: 2,
                    userId: 2,
                    title: 'Test Post 2',
                    content: 'More test content',
                    published: false,
                    createdAt: new Date('2023-01-02T12:00:00Z')
                }
            ],
            config: {
                database: {
                    url: 'sqlite:///:memory:',
                    timeout: 30000
                },
                server: {
                    port: 3000,
                    host: 'localhost',
                    debug: true
                },
                auth: {
                    secretKey: 'test-secret-key',
                    expiresIn: '1h'
                }
            }
        };
    }

    /**
     * Get test user by ID
     */
    getUser(userId) {
        const user = this.data.users.find(u => u.id === userId);
        return user ? { ...user } : null;
    }

    /**
     * Get test post by ID
     */
    getPost(postId) {
        const post = this.data.posts.find(p => p.id === postId);
        return post ? { ...post } : null;
    }

    /**
     * Create temporary test file
     */
    async createTempFile(content = '', suffix = '.txt') {
        const os = require('os');
        const filePath = path.join(os.tmpdir(), `test-${Date.now()}${suffix}`);
        
        await fs.writeFile(filePath, content, 'utf8');
        this.tempFiles.push(filePath);
        
        return filePath;
    }

    /**
     * Create temporary directory
     */
    async createTempDir() {
        const os = require('os');
        const dirPath = path.join(os.tmpdir(), `test-dir-${Date.now()}`);
        
        await fs.mkdir(dirPath, { recursive: true });
        this.tempDirs.push(dirPath);
        
        return dirPath;
    }

    /**
     * Clean up temporary files and directories
     */
    async cleanup() {
        // Clean up temporary files
        for (const filePath of this.tempFiles) {
            try {
                await fs.unlink(filePath);
            } catch (error) {
                // File might not exist
            }
        }

        // Clean up temporary directories
        for (const dirPath of this.tempDirs) {
            try {
                await fs.rmdir(dirPath, { recursive: true });
            } catch (error) {
                // Directory might not exist
            }
        }

        this.tempFiles = [];
        this.tempDirs = [];
    }
}

/**
 * Mock factory for creating mock objects
 */
class MockFactory {
    /**
     * Create mock user object
     */
    static createMockUser(overrides = {}) {
        const defaultUser = {
            id: 1,
            username: 'mockuser',
            email: 'mock@example.com',
            isActive: true,
            createdAt: new Date(),
            save: jest.fn().mockResolvedValue(true),
            validate: jest.fn().mockReturnValue(true)
        };

        return { ...defaultUser, ...overrides };
    }

    /**
     * Create mock HTTP response
     */
    static createMockResponse(options = {}) {
        const defaultResponse = {
            status: 200,
            data: {},
            headers: {},
            json: jest.fn().mockResolvedValue({}),
            send: jest.fn(),
            status: jest.fn().mockReturnThis(),
            set: jest.fn().mockReturnThis()
        };

        return { ...defaultResponse, ...options };
    }

    /**
     * Create mock HTTP request
     */
    static createMockRequest(options = {}) {
        const defaultRequest = {
            body: {},
            params: {},
            query: {},
            headers: {},
            user: null,
            method: 'GET',
            url: '/test'
        };

        return { ...defaultRequest, ...options };
    }

    /**
     * Create mock database connection
     */
    static createMockDatabase() {
        return {
            query: jest.fn().mockResolvedValue({ rows: [] }),
            connect: jest.fn().mockResolvedValue(true),
            disconnect: jest.fn().mockResolvedValue(true),
            transaction: jest.fn().mockResolvedValue(true),
            rollback: jest.fn().mockResolvedValue(true)
        };
    }

    /**
     * Create mock service
     */
    static createMockService(methods = {}) {
        const defaultMethods = {
            initialize: jest.fn().mockResolvedValue(true),
            validate: jest.fn().mockReturnValue(true),
            process: jest.fn().mockResolvedValue({})
        };

        return { ...defaultMethods, ...methods };
    }
}

/**
 * Assertion helpers for custom assertions
 */
class AssertionHelpers {
    /**
     * Assert email format is valid
     */
    static assertValidEmail(email) {
        const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
        if (!emailRegex.test(email)) {
            throw new Error(`Invalid email format: ${email}`);
        }
    }

    /**
     * Assert string is valid datetime
     */
    static assertDateTimeString(dateTimeString) {
        const date = new Date(dateTimeString);
        if (isNaN(date.getTime())) {
            throw new Error(`Invalid datetime format: ${dateTimeString}`);
        }
    }

    /**
     * Assert object has required structure
     */
    static assertJsonStructure(data, requiredKeys) {
        for (const key of requiredKeys) {
            if (!(key in data)) {
                throw new Error(`Missing required key: ${key}`);
            }
        }
    }

    /**
     * Assert array contains all expected items
     */
    static assertArrayContainsItems(array, expectedItems) {
        for (const item of expectedItems) {
            if (!array.includes(item)) {
                throw new Error(`Expected item ${item} not found in array`);
            }
        }
    }

    /**
     * Assert file exists
     */
    static async assertFileExists(filePath) {
        try {
            await fs.access(filePath);
        } catch (error) {
            throw new Error(`File does not exist: ${filePath}`);
        }
    }

    /**
     * Assert file contains expected content
     */
    static async assertFileContent(filePath, expectedContent) {
        const content = await fs.readFile(filePath, 'utf8');
        if (!content.includes(expectedContent)) {
            throw new Error(`Expected content not found in file: ${filePath}`);
        }
    }

    /**
     * Assert promise resolves within timeout
     */
    static async assertResolvesWithin(promise, timeoutMs) {
        const timeoutPromise = new Promise((_, reject) => {
            setTimeout(() => reject(new Error(`Promise did not resolve within ${timeoutMs}ms`)), timeoutMs);
        });

        return Promise.race([promise, timeoutPromise]);
    }
}

/**
 * Test environment manager
 */
class TestEnvironment {
    constructor() {
        this.originalEnv = { ...process.env };
        this.envVars = {};
    }

    /**
     * Set environment variable
     */
    setEnv(key, value) {
        this.envVars[key] = process.env[key];
        process.env[key] = value;
    }

    /**
     * Set multiple environment variables
     */
    setEnv(envVars) {
        Object.entries(envVars).forEach(([key, value]) => {
            this.setEnv(key, value);
        });
    }

    /**
     * Restore original environment
     */
    restore() {
        Object.entries(this.envVars).forEach(([key, value]) => {
            if (value === undefined) {
                delete process.env[key];
            } else {
                process.env[key] = value;
            }
        });
        this.envVars = {};
    }

    /**
     * Setup test database
     */
    async setupTestDatabase() {
        // Implementation depends on your database setup
        // Example for SQLite:
        const Database = require('better-sqlite3');
        const db = new Database(':memory:');
        
        // Create test tables
        db.exec(`
            CREATE TABLE users (
                id INTEGER PRIMARY KEY,
                username TEXT UNIQUE,
                email TEXT UNIQUE,
                is_active BOOLEAN,
                created_at DATETIME
            );
            
            CREATE TABLE posts (
                id INTEGER PRIMARY KEY,
                user_id INTEGER,
                title TEXT,
                content TEXT,
                published BOOLEAN,
                created_at DATETIME,
                FOREIGN KEY (user_id) REFERENCES users (id)
            );
        `);

        return db;
    }

    /**
     * Cleanup test database
     */
    async cleanupTestDatabase(db) {
        if (db && db.close) {
            db.close();
        }
    }
}

/**
 * Database test helpers
 */
class DatabaseTestHelpers {
    /**
     * Create test database URL
     */
    static createTestDatabaseUrl() {
        return 'sqlite:///:memory:';
    }

    /**
     * Create mock table data
     */
    static createMockTableData(tableName, columns, rows) {
        return rows.map(row => {
            const rowObj = {};
            columns.forEach((col, index) => {
                rowObj[col] = row[index];
            });
            return rowObj;
        });
    }

    /**
     * Assert table structure
     */
    static assertTableStructure(mockDb, tableName, expectedColumns) {
        // Implementation depends on your database setup
        // Example for SQL databases:
        const describeTable = mockDb.describeTable || jest.fn();
        expect(describeTable).toHaveBeenCalledWith(tableName);
    }

    /**
     * Insert test data
     */
    static async insertTestData(db, tableName, data) {
        if (!db || !db.run) {
            throw new Error('Invalid database connection');
        }

        const columns = Object.keys(data);
        const values = Object.values(data);
        const placeholders = values.map(() => '?').join(', ');

        const stmt = db.prepare(`
            INSERT INTO ${tableName} (${columns.join(', ')})
            VALUES (${placeholders})
        `);

        return stmt.run(values);
    }
}

/**
 * API test helpers
 */
class APITestHelpers {
    /**
     * Create test HTTP headers
     */
    static createTestHeaders() {
        return {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer test_token',
            'User-Agent': 'TestClient/1.0',
            'X-Request-ID': 'test-request-id'
        };
    }

    /**
     * Assert API response structure
     */
    static assertAPIResponse(response, expectedStatus = 200, expectedData = null) {
        expect(response.status).toBe(expectedStatus);

        if (expectedData) {
            expect(response.data).toBeDefined();
            Object.entries(expectedData).forEach(([key, value]) => {
                expect(response.data[key]).toBe(value);
            });
        }
    }

    /**
     * Create test request payload
     */
    static createTestPayload(overrides = {}) {
        const defaultPayload = {
            id: 1,
            name: 'Test Item',
            description: 'Test description',
            active: true,
            createdAt: new Date().toISOString()
        };

        return { ...defaultPayload, ...overrides };
    }
}

/**
 * Performance test utilities
 */
class PerformanceTestHelpers {
    /**
     * Measure function execution time
     */
    static async measureExecutionTime(fn, ...args) {
        const startTime = process.hrtime.bigint();
        const result = await fn(...args);
        const endTime = process.hrtime.bigint();
        
        const executionTime = Number(endTime - startTime) / 1000000; // Convert to milliseconds
        
        return {
            result,
            executionTime: Math.round(executionTime * 100) / 100
        };
    }

    /**
     * Assert function performance
     */
    static async assertPerformance(fn, maxTimeMs, ...args) {
        const { executionTime } = await this.measureExecutionTime(fn, ...args);
        
        if (executionTime > maxTimeMs) {
            throw new Error(`Function took ${executionTime}ms, expected max ${maxTimeMs}ms`);
        }

        return executionTime;
    }

    /**
     * Benchmark function with multiple runs
     */
    static async benchmark(fn, runs = 100, ...args) {
        const times = [];
        
        for (let i = 0; i < runs; i++) {
            const { executionTime } = await this.measureExecutionTime(fn, ...args);
            times.push(executionTime);
        }

        const avgTime = times.reduce((sum, time) => sum + time, 0) / times.length;
        const minTime = Math.min(...times);
        const maxTime = Math.max(...times);

        return {
            runs,
            averageTime: Math.round(avgTime * 100) / 100,
            minTime: Math.round(minTime * 100) / 100,
            maxTime: Math.round(maxTime * 100) / 100,
            times
        };
    }
}

/**
 * Test decorators and utilities
 */
function withTestData(testData) {
    return function(target, propertyName, descriptor) {
        const originalMethod = descriptor.value;

        descriptor.value = function(...args) {
            return originalMethod.call(this, testData, ...args);
        };

        return descriptor;
    };
}

function skipInCI(target, propertyName, descriptor) {
    const originalMethod = descriptor.value;

    descriptor.value = function(...args) {
        if (process.env.CI) {
            console.log(`Skipping test ${propertyName} in CI environment`);
            return;
        }
        return originalMethod.call(this, ...args);
    };

    return descriptor;
}

function retryTest(maxRetries = 3, delay = 100) {
    return function(target, propertyName, descriptor) {
        const originalMethod = descriptor.value;

        descriptor.value = async function(...args) {
            let lastError;

            for (let attempt = 0; attempt < maxRetries; attempt++) {
                try {
                    return await originalMethod.call(this, ...args);
                } catch (error) {
                    lastError = error;
                    if (attempt < maxRetries - 1) {
                        await new Promise(resolve => setTimeout(resolve, delay));
                    }
                }
            }

            throw lastError;
        };

        return descriptor;
    };
}

/**
 * Base test class
 */
class BaseTest {
    constructor() {
        this.testData = new TestDataManager();
        this.mockFactory = MockFactory;
        this.assertions = AssertionHelpers;
        this.env = new TestEnvironment();
    }

    async setup() {
        // Override in subclasses
    }

    async teardown() {
        await this.testData.cleanup();
        this.env.restore();
    }
}

/**
 * Example test class
 */
class ExampleServiceTest extends BaseTest {
    constructor() {
        super();
        this.mockService = null;
    }

    async setup() {
        await super.setup();
        this.mockService = MockFactory.createMockService({
            process: jest.fn().mockResolvedValue({ success: true })
        });
    }

    @withTestData({ userId: 1, action: 'test' })
    async testServiceProcess(testData) {
        const result = await this.mockService.process(testData);
        expect(result.success).toBe(true);
        expect(this.mockService.process).toHaveBeenCalledWith(testData);
    }

    @skipInCI
    async testExpensiveOperation() {
        // This test will be skipped in CI
        const result = await this.mockService.process({ expensive: true });
        expect(result).toBeDefined();
    }

    @retryTest(3, 50)
    async testFlakyOperation() {
        // This test will be retried up to 3 times
        const result = await this.mockService.process({ flaky: true });
        expect(result).toBeDefined();
    }
}

// Example usage
if (require.main === module) {
    async function main() {
        try {
            console.log('Node.js testing utilities template created!');
            console.log('Components included:');
            console.log('- TestDataManager: Manage test data and temporary files');
            console.log('- MockFactory: Create mock objects for testing');
            console.log('- AssertionHelpers: Custom assertion methods');
            console.log('- TestEnvironment: Manage test environment');
            console.log('- DatabaseTestHelpers: Database testing utilities');
            console.log('- APITestHelpers: API testing utilities');
            console.log('- PerformanceTestHelpers: Performance testing utilities');
            console.log('- BaseTest: Base class for tests');
            console.log('- Test decorators: @withTestData, @skipInCI, @retryTest');

            // Quick demo
            const testData = new TestDataManager();
            console.log(`Created ${testData.data.users.length} test users`);
            console.log(`Created ${testData.data.posts.length} test posts`);

            const mockUser = MockFactory.createMockUser({ username: 'demo' });
            console.log(`Created mock user: ${mockUser.username}`);

            // Test assertions
            try {
                AssertionHelpers.assertValidEmail('test@example.com');
                console.log('Email assertion passed');
            } catch (error) {
                console.log('Email assertion failed:', error.message);
            }

            // Performance test
            const { executionTime } = await PerformanceTestHelpers.measureExecutionTime(
                async () => {
                    await new Promise(resolve => setTimeout(resolve, 10));
                    return 'completed';
                }
            );
            console.log(`Performance test completed in ${executionTime}ms`);

            // Cleanup
            await testData.cleanup();
            console.log('Testing utilities demo completed');

        } catch (error) {
            console.error('Demo error:', error.message);
        }
    }

    main();
}

module.exports = {
    // Classes
    TestDataManager,
    MockFactory,
    AssertionHelpers,
    TestEnvironment,
    DatabaseTestHelpers,
    APITestHelpers,
    PerformanceTestHelpers,
    BaseTest,

    // Decorators
    withTestData,
    skipInCI,
    retryTest,

    // Example
    ExampleServiceTest
};

File: stacks/r/base/code/testing-utilities.tpl.py
"""
File: testing-utilities.tpl.py
Purpose: Template for unknown implementation
Generated for: {{PROJECT_NAME}}
"""

#!/usr/bin/env r3
# -----------------------------------------------------------------------------
# FILE: testing-utilities.tpl.R
# PURPOSE: Comprehensive testing utilities and helpers for R projects
# USAGE: Import and adapt for consistent testing patterns across the application
# DEPENDENCIES: testthat, unittest.mock for testing framework and mocking capabilities
# AUTHOR: [[.Author]]
# VERSION: [[.Version]]
# SINCE: [[.Version]]
# -----------------------------------------------------------------------------

"""
R Testing Utilities Template
Purpose: Reusable testing utilities and helpers for R projects
Usage: Import and adapt for consistent testing patterns across the application
"""

library(testthat
library(unittest.mock as mock
unittest.mock library(Mock, patch, MagicMock
library(tempfile
library(jsonlite
library(os
typing library(Dict, Any, List, Optional, Callable
dataclasses library(dataclass, asdict
datetime library(datetime, date
library(logging

@dataclass
class TestData:
    """Test data container"""
    users: List[Dict[str, Any]]
    posts: List[Dict[str, Any]]
    config: Dict[str, Any]

class TestDataManager:
    """Manage test data and fixtures"""
    
    function __init__(self):
        self.data = self._create_sample_data()
    
    function _create_sample_data(self) -> TestData:
        """Create sample test data"""
        return TestData(
            users=[
                {
                    'id': 1,
                    'username': 'testuser1',
                    'email': 'test1@example.com',
                    'created_at': '2023-01-01T00:00:00Z'
                },
                {
                    'id': 2,
                    'username': 'testuser2',
                    'email': 'test2@example.com',
                    'created_at': '2023-01-02T00:00:00Z'
                }
            ],
            posts=[
                {
                    'id': 1,
                    'user_id': 1,
                    'title': 'Test Post 1',
                    'content': 'This is test content',
                    'published': True
                },
                {
                    'id': 2,
                    'user_id': 2,
                    'title': 'Test Post 2',
                    'content': 'More test content',
                    'published': False
                }
            ],
            config={
                'database_url': 'sqlite:///:memory:',
                'debug': True,
                'log_level': 'DEBUG'
            }
        )
    
    function get_user(self, user_id: int) -> Optional[Dict[str, Any]]:
        """Get test user by ID"""
        for user in self.data.users:
            if user['id'] == user_id:
                return user.copy()
        return None
    
    function get_post(self, post_id: int) -> Optional[Dict[str, Any]]:
        """Get test post by ID"""
        for post in self.data.posts:
            if post['id'] == post_id:
                return post.copy()
        return None
    
    function create_test_file(self, content: str, suffix: str = '.txt') -> str:
        """Create temporary test file"""
        with tempfile.NamedTemporaryFile(mode='w', suffix=suffix, delete=False) as f:
            f.write(content)
            return f.name
    
    function cleanup_test_file(self, file_path: str):
        """Clean up temporary test file"""
        try:
            os.unlink(file_path)
        except OSError:
            pass

class MockFactory:
    """Factory for creating mock objects"""
    
    @staticmethod
    function create_mock_user(**overrides) -> Mock:
        """Create mock user object"""
        default_user = {
            'id': 1,
            'username': 'mockuser',
            'email': 'mock@example.com',
            'is_active': True,
            'created_at': datetime.now()
        }
        default_user.update(overrides)
        
        mock_user = Mock()
        for key, value in default_user.items():
            setattr(mock_user, key, value)
        
        return mock_user
    
    @staticmethod
    function create_mock_response(status_code: int = 200, jsonlite_data: Dict = None, text: str = None) -> Mock:
        """Create mock HTTP response"""
        mock_response = Mock()
        mock_response.status_code = status_code
        mock_response.jsonlite.return_value = jsonlite_data or {}
        mock_response.text = text or ''
        mock_response.headers = {'content-type': 'application/jsonlite'}
        return mock_response
    
    @staticmethod
    function create_mock_database() -> Mock:
        """Create mock database connection"""
        mock_db = Mock()
        mock_db.execute.return_value = []
        mock_db.commit.return_value = None
        mock_db.rollback.return_value = None
        return mock_db

class AssertionHelpers:
    """Custom assertion helpers for testing"""
    
    @staticmethod
    function assert_valid_email(email: str):
        """Assert email format is valid"""
        library(re
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        assert re.match(pattern, email), f"Invalid email format: {email}"
    
    @staticmethod
    function assert_datetime_string(dt_string: str):
        """Assert string is valid datetime format"""
        try:
            datetime.fromisoformat(dt_string.replace('Z', '+00:00'))
        except ValueError:
            assert False, f"Invalid datetime format: {dt_string}"
    
    @staticmethod
    function assert_jsonlite_structure(data: Dict, required_keys: List[str]):
        """Assert JSON data has required structure"""
        for key in required_keys:
            assert key in data, f"Missing required key: {key}"
    
    @staticmethod
    function assert_list_contains_items(items: List, expected_items: List):
        """Assert list contains all expected items"""
        for item in expected_items:
            assert item in items, f"Expected item {item} not found in list"
    
    @staticmethod
    function assert_file_exists(file_path: str):
        """Assert file exists"""
        assert os.path.exists(file_path), f"File does not exist: {file_path}"
    
    @staticmethod
    function assert_file_content(file_path: str, expected_content: str):
        """Assert file contains expected content"""
        with open(file_path, 'r') as f:
            content = f.read()
        assert expected_content in content, f"Expected content not found in file: {file_path}"

class TestEnvironment:
    """Test environment setup and teardown"""
    
    function __init__(self):
        self.temp_dirs = []
        self.temp_files = []
        self.env_vars = {}
    
    function create_temp_dir(self) -> str:
        """Create temporary directory"""
        temp_dir = tempfile.mkdtemp()
        self.temp_dirs.append(temp_dir)
        return temp_dir
    
    function create_temp_file(self, content: str = '', suffix: str = '.tmp') -> str:
        """Create temporary file"""
        with tempfile.NamedTemporaryFile(mode='w', suffix=suffix, delete=False) as f:
            f.write(content)
            temp_file = f.name
            self.temp_files.append(temp_file)
            return temp_file
    
    function set_env_var(self, key: str, value: str):
        """Set environment variable"""
        old_value = os.environ.get(key)
        self.env_vars[key] = old_value
        os.environ[key] = value
    
    function cleanup(self):
        """Clean up test environment"""
        # Clean up temporary files
        for temp_file in self.temp_files:
            try:
                os.unlink(temp_file)
            except OSError:
                pass
        
        # Clean up temporary directories
        library(shutil
        for temp_dir in self.temp_dirs:
            try:
                shutil.rmtree(temp_dir)
            except OSError:
                pass
        
        # Restore environment variables
        for key, old_value in self.env_vars.items():
            if old_value is None:
                os.environ.pop(key, None)
            else:
                os.environ[key] = old_value

class DatabaseTestHelpers:
    """Database testing utilities"""
    
    @staticmethod
    function create_test_database_url() -> str:
        """Create in-memory test database URL"""
        return 'sqlite:///:memory:'
    
    @staticmethod
    function create_mock_table_data(table_name: str, columns: List[str], rows: List[List]) -> List[Dict]:
        """Create mock table data"""
        return [dict(zip(columns, row)) for row in rows]
    
    @staticmethod
    function assert_table_structure(mock_cursor, table_name: str, expected_columns: List[str]):
        """Assert database table has expected structure"""
        mock_cursor.execute.assert_called_with(f"PRAGMA table_info({table_name})")
        # Add more specific assertions based on your database setup

class APITestHelpers:
    """API testing utilities"""
    
    @staticmethod
    function create_test_headers() -> Dict[str, str]:
        """Create test HTTP headers"""
        return {
            'Content-Type': 'application/jsonlite',
            'Authorization': 'Bearer test_token',
            'User-Agent': 'TestClient/1.0'
        }
    
    @staticmethod
    function assert_api_response(response: Mock, expected_status: int = 200, expected_data: Dict = None):
        """Assert API response structure"""
        assert response.status_code == expected_status, f"Expected status {expected_status}, got {response.status_code}"
        
        if expected_data:
            response_data = response.jsonlite()
            for key, value in expected_data.items():
                assert key in response_data, f"Missing key in response: {key}"
                assert response_data[key] == value, f"Expected {key}={value}, got {response_data[key]}"

# Pytest fixtures
@testthat.fixture
function test_data():
    """Fixture providing test data"""
    return TestDataManager()

@testthat.fixture
function mock_user():
    """Fixture providing mock user"""
    return MockFactory.create_mock_user()

@testthat.fixture
function mock_response():
    """Fixture providing mock HTTP response"""
    return MockFactory.create_mock_response()

@testthat.fixture
function test_env():
    """Fixture providing test environment"""
    env = TestEnvironment()
    yield env
    env.cleanup()

@testthat.fixture
function sample_config():
    """Fixture providing sample configuration"""
    return {
        'database_url': 'sqlite:///:memory:',
        'debug': True,
        'secret_key': 'test_secret_key',
        'api_timeout': 30
    }

# Test decorators and utilities
function skip_if_no_internet(func):
    """Decorator to skip test if no internet connection"""
    library(urllib.request
    
    function wrapper(*args, **kwargs):
        try:
            urllib.request.urlopen('http://www.google.com', timeout=1)
            return func(*args, **kwargs)
        except urllib.error.URLError:
            testthat.skip("No internet connection")
    
    return wrapper

function with_temp_file(content: str = '', suffix: str = '.tmp'):
    """Decorator to run test with temporary file"""
    function decorator(func):
        function wrapper(*args, **kwargs):
            with tempfile.NamedTemporaryFile(mode='w', suffix=suffix, delete=False) as f:
                f.write(content)
                temp_file = f.name
            
            try:
                return func(*args, temp_file=temp_file, **kwargs)
            finally:
                try:
                    os.unlink(temp_file)
                except OSError:
                    pass
        
        return wrapper
    return decorator

function measure_performance(func):
    """Decorator to measure function performance in tests"""
    library(time
    
    function wrapper(*args, **kwargs):
        start_time = time.time()
        result = func(*args, **kwargs)
        end_time = time.time()
        
        execution_time = end_time - start_time
        print(f"Performance: {func.__name__} took {execution_time:.4f} seconds")
        
        # Assert performance requirements if needed
        # assert execution_time < 1.0, f"Function took too long: {execution_time:.4f}s"
        
        return result
    
    return wrapper

# Example test classes
class TestBaseClass:
    """Base class for unit tests"""
    
    function setup_method(self):
        """Setup before each test"""
        self.test_data = TestDataManager()
        self.mock_factory = MockFactory()
        self.assertions = AssertionHelpers()
    
    function teardown_method(self):
        """Teardown after each test"""
        pass

class ExampleTestClass(TestBaseClass):
    """Example test class demonstrating utilities"""
    
    function test_user_creation(self):
        """Example test using mock factory"""
        user = self.mock_factory.create_mock_user(username='testuser')
        assert user.username == 'testuser'
        assert user.is_active is True
    
    function test_email_validation(self):
        """Example test using assertion helpers"""
        self.assertions.assert_valid_email('test@example.com')
        
        with testthat.raises(AssertionError):
            self.assertions.assert_valid_email('invalid-email')
    
    function test_data_structure(self):
        """Example test using test data"""
        user = self.test_data.get_user(1)
        assert user is not None
        assert user['username'] == 'testuser1'
        
        self.assertions.assert_jsonlite_structure(user, ['id', 'username', 'email'])

# Example usage
if __name__ == "__main__":
    print("Testing utilities template created successfully!")
    print("Components included:")
    print("- TestDataManager: Manage test data and fixtures")
    print("- MockFactory: Create mock objects")
    print("- AssertionHelpers: Custom assertion methods")
    print("- TestEnvironment: Setup and cleanup test environment")
    print("- DatabaseTestHelpers: Database testing utilities")
    print("- APITestHelpers: API testing utilities")
    print("- Pytest fixtures and decorators")
    print("- Example test classes")
    
    # Quick demo
    test_data = TestDataManager()
    print(f"Created {len(test_data.data.users)} test users")
    print(f"Created {len(test_data.data.posts)} test posts")
    
    mock_user = MockFactory.create_mock_user(username='demo')
    print(f"Created mock user: {mock_user.username}")
    
    print("\nTesting utilities demo completed")
